pyi.py

```diff
@@ -187,6 +239,76 @@ def in_class(self) -> bool:
         """Determine whether we are inside a `class` statement"""
         return bool(self._class_nesting)
 
+    def _check_typing_object(
+        self, node: ast.Attribute | ast.ImportFrom, module_name: str, object_name: str
+    ) -> None:
+        if object_name in _BAD_COLLECTIONS_ALIASES:
+            error_code, blacklist = Y022, _BAD_COLLECTIONS_ALIASES
+        elif object_name == "AsyncContextManager":
+            error_code = Y022
+            blacklist = {
+                "AsyncContextManager": '"contextlib.AbstractAsyncContextManager"'
+            }
+        elif module_name == "typing":
+            if object_name not in _BAD_BUILTINS_ALIASES:
+                return
+            error_code, blacklist = Y022, _BAD_BUILTINS_ALIASES
+        elif object_name in _BAD_COLLECTIONS_ABC_ALIASES:
+            error_code, blacklist = Y023, _BAD_COLLECTIONS_ABC_ALIASES
+        elif object_name in _TYPING_NOT_TYPING_EXTENSIONS:
+            error_code, blacklist = Y023, _TYPING_NOT_TYPING_EXTENSIONS
+        elif object_name == "ContextManager":
+            error_code = Y023
+            blacklist = {
+                "ContextManager": '"contextlib.AbstractContextManager" or "typing.ContextManager"'
```

Comment:
Shouldn't we pick one to recommend?

Reply:
`contextlib.AbstractContextManager` will fail on Python 2, typeshed still supports Python 2

Reply:
I see, then maybe we should say something like `'"contextlib.AbstractContextManager" (or "typing.ContextManager" in Python 2-compatible code)'`. When a user sees this error message, I want them to have enough information to know how to fix it, and if we give them two names, how are they going to know which one to use?

Reply:
That makes sense... What do you think about the `collections.abc` case, though? `use "collections.abc.Awaitable", unless you stumble across a random mypy/pytype bug, in which case use `typing.Awaitable`?

Reply:
I guess we should fix those bugs so this isn't an issue any more :) I don't recall the details of those bugs, are they common? Maybe we should just recommend typing until it's always safe to use collections.abc.

Reply:
> I guess we should fix those bugs so this isn't an issue any more :)

Indeed 😀

>  I don't recall the details of those bugs, are they common?

- https://github.com/python/mypy/issues/11860
- https://github.com/google/pytype/issues/1096

Reply:
TBF, the mypy issue looks typeshed-specific, and the pytype may be as well. But we certainly can't yet run a codemodding script using `collections.abc` everywhere in typeshed (if we could, I would have already done so ;)